File: project/sungwoong-onboarding/src/test/java/com/multi/sungwoongonboarding/responses/application/ResponseServiceTest.java
package com.multi.sungwoongonboarding.responses.application;

import com.multi.sungwoongonboarding.forms.infrastructure.FormJpaRepository;
import com.multi.sungwoongonboarding.forms.infrastructure.FormsJpaEntity;
import com.multi.sungwoongonboarding.questions.domain.Questions;
import com.multi.sungwoongonboarding.questions.infrastructure.QuestionJpaEntity;
import com.multi.sungwoongonboarding.responses.dto.AnswerCreateRequest;
import com.multi.sungwoongonboarding.responses.dto.ResponseCreateRequest;
import com.multi.sungwoongonboarding.responses.infrastructure.responses.ResponseJpaEntity;
import com.multi.sungwoongonboarding.responses.infrastructure.responses.ResponsesJpaRepository;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;


@SpringBootTest
@ActiveProfiles("test")
@EnableJpaAuditing
@Sql(scripts = "/sql/form_insert_data.sql")
@Transactional
class ResponseServiceTest {

    @Autowired
    FormJpaRepository formJpaRepository;

    @Autowired
    ResponseService responseService;

    @Autowired
    private ResponsesJpaRepository responsesJpaRepository;


    @Test
    @DisplayName("sql 파일 테스트")
    public void sqlTest() {

        //Given
        Optional<FormsJpaEntity> byId = formJpaRepository.findById(1L);

        //Expected
        Assertions.assertThat(byId.isPresent()).isTrue();
        Assertions.assertThat(byId.get().getQuestions().size()).isEqualTo(3);
        Assertions.assertThat(byId.get().getQuestions().get(0).getOptions().size()).isEqualTo(2);
    }

    @Test
    @DisplayName("ResponseService 응답지 생성 기능 - 성공")
    public void submitResponseTest() {
        // Given
        // 여기에 응답 제출을 위한 테스트 데이터를 준비합니다.
        FormsJpaEntity formsJpaEntity = formJpaRepository.findById(1L).orElseThrow(AssertionError::new);
        List<QuestionJpaEntity> questions = formsJpaEntity.getQuestions();

        ResponseCreateRequest responseCreateRequest = ResponseCreateRequest.builder()
                .formId(formsJpaEntity.getId())
                .userId("123햅버거러버123")
                .answerCreateRequests(
                        List.of(
                                AnswerCreateRequest.builder()
                                        .questionId(questions.get(0).getId())
                                        .optionId(questions.get(0).getOptions().get(0).getId())
                                        .build(),
                                AnswerCreateRequest.builder()
                                        .questionId(questions.get(1).getId())
                                        .optionId(questions.get(1).getOptions().get(0).getId())
                                        .build(),
                                AnswerCreateRequest.builder()
                                        .questionId(questions.get(1).getId())
                                        .optionId(questions.get(1).getOptions().get(1).getId())
                                        .build(),
                                AnswerCreateRequest.builder()
                                        .questionId(3L)
                                        .answerText("햄버거는 다 맛있다.")
                                        .build()
                        )).build();

        // When
        // 여기에 응답 제출 메소드를 호출합니다.
        responseService.submitResponse(responseCreateRequest);
        Optional<ResponseJpaEntity> byId = responsesJpaRepository.findById(1L);

        // Then
        // 여기에 응답 제출 결과를 검증하는 로직을 작성합니다.
        Assertions.assertThat(byId.isPresent()).isTrue();
        Assertions.assertThat(byId.get().getId()).isEqualTo(1L);
        Assertions.assertThat(byId.get().getAnswers().size()).isEqualTo(4L);
        Assertions.assertThat(byId.get().getAnswers().get(0).getOriginalQuestionType()).isEqualTo(Questions.QuestionType.SINGLE_CHOICE);
        Assertions.assertThat(byId.get().getAnswers().get(1).getOriginalQuestionType()).isEqualTo(Questions.QuestionType.MULTIPLE_CHOICE);
        Assertions.assertThat(byId.get().getAnswers().get(1).getOriginalQuestionType()).isEqualTo(Questions.QuestionType.MULTIPLE_CHOICE);
        Assertions.assertThat(byId.get().getAnswers().get(2).getOriginalQuestionType()).isEqualTo(Questions.QuestionType.MULTIPLE_CHOICE);
        Assertions.assertThat(byId.get().getAnswers().get(3).getOriginalQuestionType()).isEqualTo(Questions.QuestionType.SHORT_ANSWER);
    }
}

File: project/jihochoi-onboarding/QueryBox/src/test/kotlin/com/wlghsp/querybox/application/ResponseServiceTest.kt
package com.wlghsp.querybox.application

import com.wlghsp.querybox.domain.survey.QuestionType
import com.wlghsp.querybox.repository.ResponseRepository
import com.wlghsp.querybox.ui.dto.AnswerRequest
import com.wlghsp.querybox.ui.dto.QuestionRequest
import com.wlghsp.querybox.ui.dto.ResponseCreateRequest
import com.wlghsp.querybox.ui.dto.SurveyCreateRequest
import jakarta.transaction.Transactional
import org.assertj.core.api.AssertionsForInterfaceTypes.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
@Transactional
class ResponseServiceTest @Autowired constructor(
    private val responseService: ResponseService,
    private val surveyService: SurveyService,
    private val responseRepository: ResponseRepository,
) {

    @DisplayName("submit() — 응답 저장")
    @Test
    fun submit_savesResponse() {
        // given
        val surveyId = createSurvey()
        val question = surveyService.findSurveyWithQuestionsById(surveyId).getQuestions().first()

        val request = ResponseCreateRequest(
            answers = listOf(
                AnswerRequest(
                    questionId = question.id,
                    questionName = question.name,
                    questionType = question.type,
                    answerValue = "Kotlin",
                    selectedOptionIds = emptyList(),
                    selectedOptionTexts = emptyList()
                )
            )
        )

        // when
        responseService.submit(surveyId, request)

        // then
        val responses = responseRepository.findAllBySurveyId(surveyId)
        assertThat(responses).hasSize(1)
        assertThat(responses.first().answers!!.values().first().answerValue).isEqualTo("Kotlin")
    }

    @DisplayName("searchResponses() — 키워드 필터링")
    @Test
    fun searchResponses_filtersByKeyword() {
        // given
        val surveyId = createSurvey()
        val question = surveyService.findSurveyWithQuestionsById(surveyId).getQuestions().first()

        val kotlinRequest = ResponseCreateRequest(
            answers = listOf(
                AnswerRequest(
                    questionId = question.id,
                    questionName = question.name,
                    questionType = question.type,
                    answerValue = "Kotlin",
                    selectedOptionIds = emptyList(),
                    selectedOptionTexts = emptyList()
                )
            )
        )
        val javaRequest = kotlinRequest.copy(
            answers = listOf(
                kotlinRequest.answers.first().copy(answerValue = "Java")
            )
        )

        responseService.submit(surveyId, kotlinRequest)
        responseService.submit(surveyId, javaRequest)

        // when
        val filtered = responseService.searchResponses(
            surveyId = surveyId,
            questionKeyword = null,
            answerKeyword = "Kotlin"
        )

        // then
        assertThat(filtered).hasSize(1)
        assertThat(filtered.first().answerValue).isEqualTo("Kotlin")
    }

    private fun createSurvey(): Long =
        surveyService.create(
            SurveyCreateRequest(
                title = "응답 테스트 설문",
                description = "설명",
                questions = listOf(
                    QuestionRequest(
                        name = "선호 언어?",
                        description = "당신이 가장 선호하는 언어",
                        type = QuestionType.SHORT_TEXT,
                        required = true,
                        options = null
                    )
                )
            )
        )
}